man/cmds/rpccmd.man
' $Header: /sprite/src/cmds/rpccmd/RCS/rpccmd.man,v 1.8 92/07/10 14:59:54 kupfer Exp $ SPRITE (Berkeley)
.so \*(]ltmac.sprite
.HS RPCCMD cmds
.BS
.SH NAME
rpccmd \- Control the RPC system
.SH SYNOPSIS
\fBrpccmd\fR  [\fIoptions\fR]
.SH OPTIONS
.IP "\fB\-on\fR" 14
Turn on the server half of the RPC system.
.IP "\fB\-off\fR"
Turn off the server half of the RPC system.
.IP "\fB\-ping \fIhost\fR"
Do an RPC test against the specified host.
.IP "\fB\-b \fIsize\fR"
The amount of data to send in each RPC test.
.IP "\fB\-t \fR[\fB0|1\fR]"
Set the value of the rpc_Tracing variable.  RPC tracing facilitates
debugging, but adds overhead to each RPC. 
.IP "\fB\-D \fR[\fB0|1\fR]"
Set the value of the rpc_Debug variable.  A non-zero value causes
each RPC to be logged to \fR/dev/syslog\fR.  This really slows things down!
.IP "\fB\-T \fR[\fB0|1\fR]"
Set the value of the rpc_NoTimeouts variable.  A non-zero value
makes all RPCs try forever.  This is dangerous because a server call-back
to a dead client will hang the server.
.IP "\fB\-C \fR[\fB0|1\fR]"
Turn off/on timings of RPC on the client side.  Use 
.B rpcstat
to dump out the histograms of times that are taken by the kernel.
.IP "\fB\-S \fR[\fB0|1\fR]"
Turn off/on timing of the RPC service times as measured by the server.
.IP "\fB\-Creset\fR"
Reset the client-side timing histograms.
.IP "\fB\-Sreset\fR"
Reset the server-side timing histograms.
.IP "\fB\-maxServers \fInumber\fR"
Set the maximum number of allowed rpc server processes.
.IP "\fB\-numServers \fInumber\fR"
Create more rpc servers until this number exists.
.IP "\fB\-negAcksOn\fR"
Allow the machine to generate negative acknowledgements.
.IP "\fB\-negAcksOff\fR"
Prevent the machine from generating negative acknowledgements.
.IP "\fB\-channelNegAcksOn\fR"
Turn on the client policy of ramping down the number of channels it uses
with a server that has returned a negative acknowledgement recently.
.IP "\fB\-channelNegAcksOff\fR"
Turn off the client policy of ramping down the number of channels it uses
with a server that has returned a negative acknowledgement recently.
Use back-off instead.
.IP "\fB\-numPings\fR"
Number of pings to send to the specified host (defaults to 100).
.IP "\fB\-sanity \fR[\fB0|1\fR]"
Turn off/on sanity checking of RPC packets.  The types of sanity checks done
are intended for debugging
purposes and probably shouldn't be used during normal operation.
.BE
.SH INTRODUCTION
.PP
This command is used to set various control parameters of the RPC system.
The most important is the \fB\-on\fR option, which is used during the boot
sequence to enable RPC service after local initialization (i.e. disk checks)
is done.  The \fB\-ping\fR option is used to ping another sprite host.  This
is useful to find out if the host is up, and the make sure that the
host knows you are alive.  This does an RPC_SEND, and the amount of
data sent can be specified with the \fB\-b\fR option.
.PP
Each of the \fB\-t\fR, \fB\-D\fR, \fB\-T\fR, \fB\-C\fR, and \fB\-S\fR
options set the value
of a kernel variable.  \fB\-t\fR is used to turn on or off the trace of the last
100 RPC packets.  \fB\-D\fR is used to turn on or off debugging print statements.
\fB\-T\fR is used to turn on or off "no timeout mode".  This mode is useful
when debugging a client and you don't want the server to timeout
and think the client has crashed.  It is dangerous, however, because
if the server ever makes a call-back to a truely dead client it
will hang until the no timeout mode is turned off.  
.PP
The \fB\-C\fR and \fB\-S\fR
options turn on or off timings of individual RPCs.  \fB\-C\fR controls timing
on the client side, and \fB\-S\fR controls timing on the server side.
The timings are
kept in histogram form.  Use the 
.B \-Creset
and
.B \-Sreset
options to zero out the histograms.  Use 
.B rpcstat 
to display them.  
.SH "SEE ALSO"
rpcstat
.SH KEYWORDS
rpc, trace, debug


src/machserver/src/cmds/rpccmd/rpccmd.c
/* 
 * rpccmd.c --
 *
 *	User interface to RPC related commands of the Fs_Command system call.
 *
 * Copyright (C) 1986 Regents of the University of California
 * All rights reserved.
 * Permission to use, copy, modify, and distribute this
 * software and its documentation for any purpose and without
 * fee is hereby granted, provided that the above copyright
 * notice appear in all copies.  The University of California
 * makes no representations about the suitability of this
 * software for any purpose.  It is provided "as is" without
 * express or implied warranty.
 */

#ifndef lint
static char rcsid[] = "$Header: /user5/kupfer/spriteserver/src/cmds/rpccmd/RCS/rpccmd.c,v 1.2 92/07/13 12:59:25 kupfer Exp $ SPRITE (Berkeley)";
#endif not lint

#include <sprite.h>
#include <status.h>
#include <option.h>
#include <stdio.h>
#include <fs.h>
#include <fsCmd.h>
#include <sysStats.h>
#include <rpc.h>
#include <host.h>

/*
 * Command line options.
 */

int	rpcEnable = -1;
int	rpcDebug = -1;
int	rpcTracing = -1;
int	rpc_NoTimeouts = -1;
int	rpcClientHist = -1;
int	rpcServerHist = -1;
Boolean	resetClientHist = FALSE;
Boolean	resetServerHist = FALSE;
char	*rpcEchoHost = (char *)0;
int	blockSize = 32;
int	rpcMaxServers = -1;
int	rpcNumServers = -1;
int	rpcNegAcks = -1;
int	rpcChannelNegAcks = -1;
int	rpcNumNackBufs = -1;
int	numEchoes = 100;
int	sanityChecks = -1;

Option optionArray[] = {
    {OPT_TRUE, "on", (Address) &rpcEnable, 
	"\tAllow servicing of RPC requests."},
    {OPT_FALSE, "off", (Address) &rpcEnable, 
	"\tTurn off servicing of RPC requests."},
    {OPT_STRING, "ping", (Address) &rpcEchoHost, 
	"\tRPC test against the specified host."},
    {OPT_INT, "b", (Address) &blockSize, 
	"\tBlock size to send in RPC test."},
    {OPT_INT, "t", (Address) &rpcTracing, 
	"\tSet rpc tracing flag."},
    {OPT_INT, "T", (Address) &rpc_NoTimeouts, 
	"\tSet rpc no timeouts flag."},
    {OPT_INT, "D", (Address) &rpcDebug, 
	"\tSet rpc debug flag."},
    {OPT_INT, "C", (Address) &rpcClientHist, 
	"\tSet client histogram flag."},
    {OPT_INT, "S", (Address) &rpcServerHist, 
	"\tSet server histogram flag."},
    {OPT_TRUE, "Creset", (Address) &resetClientHist, 
	"\tReset client RPC histograms."},
    {OPT_TRUE, "Sreset", (Address) &resetServerHist, 
	"\tReset server RPC histograms."},
    {OPT_INT, "maxServers", (Address) &rpcMaxServers,
	"\tSet the maximum number of allowed rpc server processes."},
    {OPT_INT, "numServers", (Address) &rpcNumServers,
	"\tCreate more rpc servers until this number exists."},
    {OPT_TRUE, "negAcksOn", (Address) &rpcNegAcks,
	"\tTurn on negative acknowledgements on a server."},
    {OPT_FALSE, "negAcksOff", (Address) &rpcNegAcks,
	"\tTurn off negative acknowledgements on a server (default)."},
    {OPT_INT, "numNackBufs", (Address) &rpcNumNackBufs,
	"\tMake sure this number of negative ack buffers exists."},
    {OPT_TRUE, "channelNegAcksOn", (Address) &rpcChannelNegAcks,
	"\tSet the client policy for handling negative acks to ramp down number of channels."},
    {OPT_FALSE, "channelNegAcksOff", (Address) &rpcChannelNegAcks,
	"\tSet the client policy for handling negative acks to the default backoff policy."},
    {OPT_INT, "numPings", (Address) &numEchoes,
	"\tNumber of pings to send."},
    {OPT_INT, "sanity", (Address)&sanityChecks,
	"\tTurn off/on rpc sanity checks."},

};
int numOptions = sizeof(optionArray) / sizeof(Option);

/* forward references: */

static void ResetClientHist();
static void ResetServerHist();
ReturnStatus SetFlag();


/*
 *----------------------------------------------------------------------
 *
 * main --
 *
 *	Collects arguments and branch to the code for the fs command.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	Calls Fs_Command...
 *
 *----------------------------------------------------------------------
 */
main(argc, argv)
    int argc;
    char *argv[];
{
    register ReturnStatus status = SUCCESS;	/* status of system calls */

    argc = Opt_Parse(argc, argv, optionArray, numOptions, 0);

    /*
     * Set various rpc system flags.  The Fs_Command system call returns
     * the old value of the flag in place of the value passed in.
     */
    if (rpcEnable != -1) {
	status = Sys_Stats(SYS_RPC_ENABLE_SERVICE, rpcEnable, NIL);
    }
    if (rpcTracing != -1) {
	status = SetFlag(FS_SET_RPC_TRACING, rpcTracing, "RPC tracing");
    }
    if (rpc_NoTimeouts != -1) {
	status = SetFlag(FS_SET_RPC_NO_TIMEOUTS, rpc_NoTimeouts,
					    "No RPC timeouts");
    }
    if (rpcDebug != -1) {
	status = SetFlag(FS_SET_RPC_DEBUG, rpcDebug, "RPC debug prints");
    }
    if (rpcClientHist != -1) {
	status = SetFlag(FS_SET_RPC_CLIENT_HIST, rpcClientHist,
					     "Client RPC timing histograms");
    }
    if (rpcServerHist != -1) {
	status = SetFlag(FS_SET_RPC_SERVER_HIST, rpcServerHist,
					    "Server RPC timing histograms");
    }
    if (resetClientHist) {
	ResetClientHist();
    }
    if (resetServerHist) {
	ResetServerHist();
    }
    if (rpcMaxServers != -1) {
	status = Sys_Stats(SYS_RPC_SET_MAX, rpcMaxServers, NIL);
    }
    if (rpcNumServers != -1) {
	status = Sys_Stats(SYS_RPC_SET_NUM, rpcNumServers, NIL);
    }
    if (rpcNegAcks != -1) {
	status = Sys_Stats(SYS_RPC_NEG_ACKS, rpcNegAcks, NIL);
    }
    if (rpcNumNackBufs != -1) {
	status = Sys_Stats(SYS_RPC_NUM_NACK_BUFS, rpcNumNackBufs, NIL);
    }
    if (rpcChannelNegAcks != -1) {
	status = Sys_Stats(SYS_RPC_CHANNEL_NEG_ACKS, rpcChannelNegAcks, NIL);
    }
    if (rpcEchoHost != (char *)0) {
	status = RpcEcho(rpcEchoHost);
    }
    if (sanityChecks != -1) {
	status = Sys_Stats(SYS_RPC_SANITY_CHECK, sanityChecks, NIL);
    }
    exit(status);
}

RpcEcho(server)
    char *server;
{
    Host_Entry *entryPtr;
    int serverID;
    int myID;
    int error;
    char buffer[16 * 1024];
    Time deltaTime;

    entryPtr = Host_ByName(server);
    if (entryPtr == (Host_Entry *) NULL) {
	fprintf(stderr, "Unable to get host number for host '%s'\n",
		       server);
	exit(1);
    }
    serverID = entryPtr->id;

    error = Proc_GetHostIDs((int *) NULL, &myID);
    if (error != SUCCESS) {
	perror( "Proc_GetHostIDs");
	exit(error);
    }
    if (myID == serverID) {
	fprintf(stderr, "Unable to send RPC to yourself.\n");
	exit(1);
    }
    error = DoTest(TEST_RPC_SEND, serverID, numEchoes, blockSize, buffer,
	    buffer, &deltaTime);
    if (error != SUCCESS) {
	Stat_PrintMsg(error, "RPC failed:");
    } else {
	printf("Send %d bytes %d.%06d sec\n", blockSize, deltaTime.seconds,
		deltaTime.microseconds);
    }
    return(error);
}
ReturnStatus
DoTest(command, serverID, numEchoes, size, inDataBuffer, outDataBuffer, timePtr)
    int command;
    int serverID;
    int numEchoes;
    int size;
    Address inDataBuffer;
    Address outDataBuffer;
    Time *timePtr;
{
    Rpc_EchoArgs echoArgs;
    ReturnStatus error;

    echoArgs.serverID = serverID;
    echoArgs.n = numEchoes;
    echoArgs.size = size;
    echoArgs.inDataPtr = inDataBuffer;
    echoArgs.outDataPtr = outDataBuffer;
    echoArgs.deltaTimePtr = timePtr;
    error = Test_Rpc(command, &echoArgs);

    return(error);
}

ReturnStatus
SetFlag(command, value, comment)
    int command;		/* Argument to Fs_Command */
    int value;			/* Value for flag */
    char *comment;		/* For Io_Print */
{
    register int newValue;
    register ReturnStatus status;

    newValue = value;
    status = Fs_Command(command, sizeof(int), (Address) &value);
    printf("%s %s, was %s\n", comment,
		     newValue ? "on" : "off",
		     value ? "on" : "off");
    return(status);
}


/*
 *----------------------------------------------------------------------
 *
 * ResetClientHist --
 *
 *	Reset the client-side histograms.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

static void
ResetClientHist()
{
    ReturnStatus status;

    status = Sys_Stats(SYS_RPC_CLIENT_HIST, 0, NULL);
    if (status != SUCCESS) {
	fprintf(stderr, "Can't reset client RPC histograms: %s\n",
		Stat_GetMsg(status));
    }
}


/*
 *----------------------------------------------------------------------
 *
 * ResetServerHist --
 *
 *	Reset the server-side histograms.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

static void
ResetServerHist()
{
    ReturnStatus status;

    status = Sys_Stats(SYS_RPC_SERVER_HIST, 0, NULL);
    if (status != SUCCESS) {
	fprintf(stderr, "Can't reset server RPC histograms: %s\n",
		Stat_GetMsg(status));
    }
}
